### DISCUSSION TITLE
Reading from HFM when buttons are pressed

### DISCUSSION THREAD

**Discussion by monalos** [cedricp/ddt4all#1049] [Q&A]
Hello, I am trying to do a job by reading the CAN bus of my Clio IV car. I read from the can bus that the buttons of my key remote have been pressed. I cannot read on the can bus that these buttons are pressed when the car starts. Most cars deactivate the control buttons when they start, locking and unlocking do not work. When I control the HFM ecu from ddt4all, there is a BUTTON menu and I can see that the buttons are pressed there. When I check it while the car is running, I can still see that the buttons are pressed in the HFM ECU. HFM tells the can bus that the buttons are pressed before the car starts, but when the car starts, it keeps it only to itself. What I want to ask now is how can I get the part where those buttons are pressed from ddt4all using the C programming language. If I can read the buttons pressed from there, I want to do something with the software I wrote.

**Comment by matzel82** [comment]
HFM is the keyless entry and mobilphone charger/ ecu

**Reply by monalos** [reply]
Yes I know. Yes I know. what about the result?

**Reply by matzel82** [reply]
You have to choose the keyless remote ecu, not HFM. Or the UCH.

**Reply by monalos** [reply]
We are talking about very different topics. I mean, I can already see that I pressed the buttons in the RKE menu in HFM. My goal is how to access those ddt4all codes and adapt them to my own software. When the car is not running, I can see the button statuses when I connect a can bus reader. But when the car starts, I can only see the buttons on the RKE on HFM. When the car starts, the buttons cannot be pressed on the can bus line. Let me simply say that when the car is running, when you press the lock and unlock buttons on the remote control, the buttons work. do you have any idea? Does your car lock or unlock with the remote control while it's running? Generally, none of Renault's cars lock and unlock with the remote control while running, but HFM hides this.

**Reply by shrlnm** [reply]
Hello @monalos. Could you provide more detail? Which ECU (xml file) do you have? Which command do you use for reading the state of buttons, оr which CANid do you listen for it? Which filter do you use if you are listen for system frames?

**Reply by monalos** [reply]
@shrlnm   I don't have XLM files. I connect to the car's can bud line with ESP32 and MCP2515 and use the canhacker program. I see all the can bus information in can hacker. HFM can ID number is 303, if you filter 303, you can read the control buttons. When the car is not running, 303 gives can bus information, but when the car is running, there is no information from 303. Because when the car starts, HFM stores the key information and does not transmit it to the can bus. My goal is, how can I learn to press the control buttons when the car starts?

**Reply by monalos** [reply]
When I select the HFM ecus with DDT4ALL, I can see which button of the remote control I pressed in the RKE section. I can also see it when the car starts. I still haven't figured out how to see it from Canbus.

**Reply by shrlnm** [reply]
@monalos Probably I did not understand you correct but there is only "start button" information  in frame 303
```
<Comment><![CDATA[100]]></Comment>
<Sent>
<SentBytes>0303</SentBytes>
</Sent>
<Received MinBytes="7">
<DataItem Name="HFM_Refuse_to_Sleep" Ref="1" FirstByte="1"/>
<DataItem Name="StartingRequest" Ref="1" FirstByte="1" BitOffset="2"/>
<DataItem Name="IgnitionRequest" Ref="1" FirstByte="1" BitOffset="4"/>
<DataItem Name="BatTempoRequest" Ref="1" FirstByte="1" BitOffset="6"/>
<DataItem Name="HandFreeCode" Ref="1" FirstByte="2"/>
<DataItem Name="HFMRequest" Ref="1" FirstByte="4"/>
<DataItem Name="CardPresenceStatus" Ref="1" FirstByte="4" BitOffset="4"/>
<DataItem Name="IKeyCardDetectionStatus" Ref="1" FirstByte="4" BitOffset="6"/>
<DataItem Name="EmergencyEngineStop" Ref="1" FirstByte="5"/>
<DataItem Name="UserIdent" Ref="1" FirstByte="5" BitOffset="4"/>
<DataItem Name="StartButtonPushed" Ref="1" FirstByte="6"/>
<DataItem Name="SwitchOffSESDisturbers" Ref="1" FirstByte="6" BitOffset="3"/>
<DataItem Name="DriverSeatStatus_HFM" Ref="1" FirstByte="6" BitOffset="4"/>
<DataItem Name="IgnitionControlState_HFM" Ref="1" FirstByte="6" BitOffset="6"/>
<DataItem Name="RESFeedback" Ref="1" FirstByte="6" BitOffset="7"/>
</Received>
</Request>
<Request Name="HFM_to_DTOOL">

```

Anyway you may getting the status of buttons by the  request to the ECU just like ddt4all  do it. I suppose that it is not harmful to request an ECU 3-5 times per second

**Reply by monalos** [reply]
I read the status of the 3rd data button at ID 0x303 on the can bus. ID: 0x303 dlc:7 data: 0x00 0x00 0x24. In this can bus information, 0x24 is normally 0x14. When I press the door open button on the remote control, it becomes 0x24, the doors open and it becomes 0x14 again. . Now, how can I add codes like you wrote to my own software with ddt4all? When the car is not running, I read the button pressing status with my software. Can you give me an example of how I can read from ddt4all while the car is running? @shrlnm

**Reply by monalos** [reply]
Let me tell you about the software and electronic circuit I made. If I press the trunk release button on my remote control twice and hold it for the third time, the car starts remotely. When remote start is active, if the brakes are pressed or one of the doors is opened, the car stalls. The car stalls itself 10 minutes after remote start starts. Everything is great, but in order to stop the car remotely before 10 minutes, I need to read the remote control keys and include them in my software.

@shrlnm

**Reply by monalos** [reply]
When I connect to the HFM control unit with ddt4all, can I find something in the ecu to read the remote control keys? So, when the car starts, all you have to do is lock or unlock the car with the remote control; I will handle the rest with my software.

@shrlnm

**Reply by shrlnm** [reply]
@monalos You wrote above "When I control the HFM ecu from ddt4all, there is a BUTTON menu and I can see that the buttons are pressed there."
I suppose you are used xml T4_HFM_DDT2000_SW4_1_5_20111003T095811 or similar. 
Write a log file during ddt4all session and then look commands it using for reading interesting data.
Your MCU have to do the same command sequence
1) For opening session the MCU send can frame "799 3 02 10 C0"
2) periodically MCU sending command to read data "799 3 02 21 31"
3) MCU listen to responses from CANid 7B9 for results

**Reply by monalos** [reply]
![IMG_7688](https://github.com/cedricp/ddt4all/assets/76909804/4502a252-825c-47ff-a5ee-699aa2eda26b)


T4 HFM SW 14.45

I listened to Can ID 7B9 on the CAN bus line, but I couldn't see it. 7B9 is not sending any messages. The Rke page in HFM that I have added in the image shows me correctly which button was pressed when I press the remote control buttons while the car is running, as in the image. Now how and where can I get this as data? I haven't dealt with xml files so I didn't understand what you were saying.

**Reply by monalos** [reply]
![IMG_7702](https://github.com/cedricp/ddt4all/assets/76909804/3a98cce6-f116-4ad6-9f65-f1fda2549a19)

The some messages I can read from the  can bus line are as in the image.

**Reply by shrlnm** [reply]
@monalos Before listening anything on 7B9 you have to send the command to 799 first. Look at the ddt4all logs.

**Reply by monalos** [reply]
@shrlnm 
I don't know about ddt4all logs, how can I do it?

As far as I understand, I will send 02 10 C0 to dlc 3 of 799 and listen to 7B9, is it true?

I will write the command in the places in the image and send it to Can Busa.

![IMG_7723](https://github.com/cedricp/ddt4all/assets/76909804/03107e41-1b62-46b2-817c-9ed221afd522)

**Reply by shrlnm** [reply]
@monalos 
Yes
10C0 is a command for opening diagnostic session and 2131 is a command for reading data.

In ddt4all  enable logging. The logs are saving in "logs" folder

**Reply by monalos** [reply]
@shrlnm 

“”Enable logging in Ddt4all. Logs are saved in the "logs" folder “”

Can you send me an image? Where is the save part in ddt4all? 

Thank you, we will reach the result.

**Reply by shrlnm** [reply]
It writes logs unconditionally. Just look at ./logs/elm_*

**Reply by monalos** [reply]
@shrlnm 

Car obd2 connected, As you said, I sent the codes to 799, but no information came from 7B9? When I sent the 799a codes, I saw other can IDs, but I did not see button presses in any of them. I connected to ddt4all and immediately opened the rke menu and pressed the buttons and saw that the buttons were pressed as true, as in the previous images I sent you. I quickly pressed the buttons 10-15 times and added the logs. Can you comment the logs?

There are 4 buttons on my remote control, I pressed all 4 buttons one by one.



[logshfm.zip](https://github.com/cedricp/ddt4all/files/14640350/logshfm.zip)

**Reply by monalos** [reply]
@shrlnm 

Ddt4all logs ı found , I think I found it. When I send 799 3 02 21 31 and I press a button, the response comes like this: >[20:07:18.043]022131 <[0.072]022131 
105C6131—40—0202E2 

If I press another key 
>[20:07:18.043]022131 <[0.072]022131 
105C6131—80—0202E2 

3. If I press the button 
>[20:07:18.043]022131 <[0.072]022131 
   105C6131—10—0202E2 

   4. If I press the button 
   >[20:07:18.043]022131 <[0.072]022131
 105C6131—20—0202E2 

Such an answer comes.
 The 10-20-40-80 data in the middle of the incoming data comes when I press one of the 4 keys. 
 Separate data comes for each key button. I will connect it to the car and try again. I separated the changed data with ----- so that you can see.

I connected to the car via can bus via obd2, I sent the command, but the 7B9 ID does not appear. I have a spare HFM, when I try it on the table at home, it shows 7B9, but it doesn't show up in the car. Should I connect it directly to the can bus cables of the HFM in the car, rather than via OBD2? In your opinion, is it possible?

**Reply by shrlnm** [reply]
@monalos I do not know how CANhacker works. I prefer to use my own tool bus_monitor included to pyren3. I suppose that the canhacker may just miss 7XX frames as so they have lowest priority on the bus and under real load on a real car it may skip them. Try to use filters.

Also I suggest to read about the CAN protocol, how commands are transmitted thru the CAN bus. You may start with my article https://www.drive2.ru/b/472129194529128744/

**Reply by monalos** [reply]
I changed the filter to 7B9 and then no information came. "can hacker" had not been able to catch another ID before. I will try with another device.

**Comment by matzel82** [comment]
You want to make your own SW….okay…im out
